### DISCUSSION TITLE
UQL how to only get latest value by date

### DISCUSSION THREAD

**Discussion by deltamelter** [grafana/grafana-infinity-datasource#266] [Q&A]
I have a json data set with dozens of values updated (mostly) daily.  every day's data is kept for 30days.
so if I want to know the current values I have to somehow filter by date and then keep the name and the latest price for each name.  Maybe the other values for that name too, but really only intereseted in price right now.
Thought I could use UQL to summarize by date, but that doesn't give me an option to get just the single latest price, eg
```
parse-json
| summarize "date"=max("date") by "name"
| project "price", "name"
| extend "price"=tonumber("price")
``` 

Also considered order by and limit, but there are new "name" units added all the time so it is not possible to say how may to limit.

Really I just want to be able to look up the latest price value for a given name, but such filtering doesn't seem part of the datasource plugin.. ??

do I miss something?  any hints?  it there another way? I thought UQL would be most flexible.

I will attach the data set sample with just 4 such item types and 2 dates for each....
[sample.txt](https://github.com/yesoreyeram/grafana-infinity-datasource/files/8116967/sample.txt)

```json
[
    {
        "date": "2022-02-19",
        "name": "Chia",
        "price": "73.09352567",
        "volumeUsd": "6673435.21",
        "marketcap": 232291225
    },
    {
        "date": "2022-02-19",
        "name": "Spacebucks",
        "price": "0.00033132",
        "volumeUsd": "11.60",
        "marketcap": 662640
    },
    {
        "date": "2022-02-19",
        "name": "Skynet",
        "price": "0.00099721",
        "volumeUsd": "0.04",
        "marketcap": 2770
    },
    {
        "date": "2022-02-20",
        "name": "Chia",
        "price": "75.30489199",
        "volumeUsd": "17429875.72",
        "marketcap": 240015065
    },
    {
        "date": "2022-02-20",
        "name": "Spacebucks",
        "price": "0.00040157",
        "volumeUsd": "1.30",
        "marketcap": 803140
    },
    {
        "date": "2022-02-20",
        "name": "STAI",
        "price": "0.04665859",
        "volumeUsd": "166.63",
        "marketcap": 158138
    },
    {
        "date": "2022-02-21",
        "name": "Chia",
        "price": "75.230489199",
        "volumeUsd": "17429875.72",
        "marketcap": 240015065
    },
    {
        "date": "2022-02-21",
        "name": "Spacebucks",
        "price": "0.000240157",
        "volumeUsd": "1.30",
        "marketcap": 803140
    },
    {
        "date": "2022-02-21",
        "name": "STAI",
        "price": "0.024665859",
        "volumeUsd": "166.63",
        "marketcap": 158138
    }
]

```

**Comment by yesoreyeram** [comment]
with `order by "" desc` and `limit 1` you should be able to get what you want. Can you post the sample dataset?
If you don't want "name" in your results, you can use `project-away "name"` to remove that column from results.

**Reply by deltamelter** [reply]
from this sample, I expect 4 results and would like name and price only for the latest date

**Reply by deltamelter** [reply]
in reality, there would be different number of entries each day, because not every value is updated daily.  Also new names are added over time so limit can not work as I don't know how many entries

**Reply by deltamelter** [reply]
modified the sample to include 4 unique names but not all updated every day over 3 days.  I want the 4 name and only the latest price

**Reply by yesoreyeram** [reply]
I got the requirement.. Do you mind creating an issue at https://github.com/yesoreyeram/uql

**Reply by deltamelter** [reply]
sure, I'll do it a bit later...

**Reply by yesoreyeram** [reply]
With the recent 0.8.3 we have a new summarisation functions called `first`/`last`. So hopefully you can [achieve](https://grafana-infinity-datasource.herokuapp.com/explore?orgId=1&left=%7B%22datasource%22:%22Infinity%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22type%22:%22uql%22,%22source%22:%22inline%22,%22format%22:%22table%22,%22url%22:%22%22,%22url_options%22:%7B%22method%22:%22GET%22,%22data%22:%22%22%7D,%22root_selector%22:%22%22,%22columns%22:%5B%5D,%22filters%22:%5B%5D,%22global_query_id%22:%22%22,%22uql%22:%22parse-json%5Cn%7C%20order%20by%20%5C%22date%5C%22%20desc%5Cn%7C%20summarize%20%5C%22price%20value%5C%22%3Dlast(%5C%22price%5C%22)%20by%20%5C%22name%5C%22%22,%22data%22:%22%5B%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-19%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Chia%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%2273.09352567%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%226673435.21%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%20232291225%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-19%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Spacebucks%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%220.00033132%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%2211.60%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%20662640%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-19%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Skynet%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%220.00099721%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%220.04%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%202770%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-20%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Chia%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%2275.30489199%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%2217429875.72%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%20240015065%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-20%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Spacebucks%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%220.00040157%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%221.30%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%20803140%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-20%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22STAI%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%220.04665859%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%22166.63%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%20158138%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-21%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Chia%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%2275.230489199%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%2217429875.72%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%20240015065%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-21%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22Spacebucks%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%220.000240157%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%221.30%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%20803140%5Cn%20%20%20%20%7D,%5Cn%20%20%20%20%7B%5Cn%20%20%20%20%20%20%20%20%5C%22date%5C%22:%20%5C%222022-02-21%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22name%5C%22:%20%5C%22STAI%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22price%5C%22:%20%5C%220.024665859%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22volumeUsd%5C%22:%20%5C%22166.63%5C%22,%5Cn%20%20%20%20%20%20%20%20%5C%22marketcap%5C%22:%20158138%5Cn%20%20%20%20%7D%5Cn%5D%22%7D%5D,%22range%22:%7B%22from%22:%22now-1h%22,%22to%22:%22now%22%7D%7D) using the below query. 

```kusto
parse-json
| order by "date" desc
| summarize "price value"=last("price") by "name"
```

<img width="1721" alt="image" src="https://user-images.githubusercontent.com/153843/155319099-c999dcfd-4acd-425b-984e-72bff61fb4dc.png">

**Reply by deltamelter** [reply]
This works perfectly (using asc and last), thanks.
Will I have to build, or you will publish this version soon?

Thanks!!

**Reply by yesoreyeram** [reply]
It is available in https://github.com/yesoreyeram/grafana-infinity-datasource/releases/tag/v0.8.3 to download

**Reply by deltamelter** [reply]
doesn't show up when I try to install from zip, seems ok via plugin catalog, but that's only at 8.0

**Reply by yesoreyeram** [reply]
have your tried the installation methods provided here https://sriramajeyam.com/grafana-infinity-datasource/wiki/installation ?

**Reply by deltamelter** [reply]
tried several different ways.  but the way I run grafana in a docker leaves it not installed/visible without restart

**Reply by deltamelter** [reply]
✔ Downloaded yesoreyeram-infinity-datasource v0.8.3 zip successfully

Please restart Grafana after installing plugins. Refer to Grafana documentation for instructions if necessary
in the docker logs

**Reply by deltamelter** [reply]
But still doesn't work.  Only installing thru the grafana plugin catalog works, but that is only 8.0

**Reply by yesoreyeram** [reply]
how your are installing the plugin? The demo grafana instance of the plugin [installs](https://github.com/yesoreyeram/grafana-infinity-datasource/blob/main/heroku.Dockerfile#L2) from zip file. So it should work

**Reply by deltamelter** [reply]
docker run with
```-e "GF_INSTALL_PLUGINS=https://github.com/yesoreyeram/grafana-infinity-datasource/releases/download/v0.8.3/yesoreyeram-infinity-datasource-0.8.3.zip;yesoreyeram-infinity-datasource"``` using grafana/grafana:latest

**Reply by deltamelter** [reply]
I'll just wait until you publish to the grafana catalog...

**Reply by deltamelter** [reply]
I don't know what is going on with this plugin, it says it is upgraded to 0.83, but UQL just does not work beyond parse-json
![image](https://user-images.githubusercontent.com/27061628/158025885-ff0a8990-e563-4afb-9dc4-fee2c15117f5.png)
is there some additional dependency for UQL?
